pallets/foreign-investments/src/impls/mod.rs

```diff
@@ -0,0 +1,1208 @@
+				Self::notify_executed_collect_redeem(
+					who,
+					investment_id,
+					done_swap.currency_in,
+					CollectedAmount {
+						amount_collected: done_swap.amount,
+						amount_payment: amount_payment_tranche_tokens,
+					},
+				)?;
+				CollectedRedemption::<T>::remove(who, investment_id);
+				Ok(())
+			}
+			_ => Ok(()),
+		}
+		.map_err(|e: DispatchError| e)?;
+
+		// Update state iff the state includes `SwapIntoForeignDone` without
+		// `ActiveSwapIntoForeignCurrency`
+		match state {
+			RedeemState::SwapIntoForeignDone { .. } => {
+				RedemptionState::<T>::remove(who, investment_id);
+				Ok(Some(RedeemState::NoState))
+			}
+			RedeemState::RedeemingAndSwapIntoForeignDone { redeem_amount, .. } => {
+				let new_state = RedeemState::Redeeming { redeem_amount };
+				RedemptionState::<T>::insert(who, investment_id, new_state);
+				Ok(Some(new_state))
+			}
+			_ => Ok(None),
+		}
+	}
+
+	/// Updates or kills a token swap order. If the final swap amount is zero,
+	/// kills the swap order and all associated storage. Else, creates or
+	/// updates an existing swap order.
+	///
+	/// If the provided reason does not match the latest one stored in
+	/// `ForeignInvestmentInfo`, also resolves the _merge conflict_ resulting
+	/// from updating and thus overwriting opposite swaps. See
+	/// [Self::handle_concurrent_swap_orders] for details. If this results in
+	/// either an altered invest state and/or an altered redeem state, the
+	/// corresponding storage is updated and the new states returned. The latter
+	/// is required for emitting events.
+	///
+	/// NOTE: Must not call any other swap order updating function.
+	#[allow(clippy::type_complexity)]
+	#[transactional]
+	fn handle_swap_order(
+		who: &T::AccountId,
+		investment_id: T::InvestmentId,
+		maybe_swap: Option<SwapOf<T>>,
+		reason: TokenSwapReason,
+	) -> Result<
+		(
+			Option<InvestState<T>>,
+			Option<RedeemState<T::Balance, T::CurrencyId>>,
+		),
+		DispatchError,
+	> {
+		// check for concurrent conflicting swap orders
+		if TokenSwapOrderIds::<T>::get(who, investment_id).is_some() {
+			let (maybe_updated_swap, maybe_invest_state, maybe_redeem_state, swap_reason) =
+				Self::handle_concurrent_swap_orders(who, investment_id)?;
+
+			// Update or kill swap order with updated order having priority in case it was
+			// overwritten
+			if let Some(swap_order) = maybe_updated_swap {
+				Self::place_swap_order(who, investment_id, swap_order, swap_reason)?;
+			} else {
+				Self::kill_swap_order(who, investment_id)?;
+			}
+
+			// Update invest state and kill if NoState
+			InvestmentState::<T>::mutate_exists(who, investment_id, |current_invest_state| {
+				match &maybe_invest_state {
+					Some(state) if state != &InvestState::NoState => {
+						*current_invest_state = Some(state.clone());
+					}
+					Some(state) if state == &InvestState::NoState => {
+						*current_invest_state = None;
+					}
+					_ => (),
+				}
+			});
```

Comment:
**MAYBE-CR**

Shouldn't we also return the updated state here, like for redeeming, instead `maybe_invest_state` that is actually used further down?

Reply:
We are returning the updated invest state which is `maybe_invest_state` as a result of executing `Self::handle_concurrent_swap_orders`. For the `RedeemState` however, we might need to transition the `maybe_redeem_state` returned by `handle_concurrent_swap_orders` in case a post-redeem-collecting-swap is done. This is the case iff the `RedeemState` includes `SwapIntoForeignDone` without `ActiceSwapIntoForeignCurrency`, see the comment: https://github.com/centrifuge/centrifuge-chain/blob/feat/wf-foreign-investments/pallets/foreign-investments/src/impls/mod.rs#L689-L709